keymanapp/keyman
# feat(linux): Add dbus system service

This implements a dbus system service that allows to provide the missing functionality when running under Wayland.

Fixes #4273.

# User Testing

## Preparations

- The tests should be run on these Linux platforms:
  - **GROUP_JAMMY_X11**: Ubuntu 22.04 Jammy with Gnome Shell and X11
  - **GROUP_JAMMY_WAYLAND**: Ubuntu 22.04 Jammy with Gnome Shell and Wayland
  - **GROUP_LUNAR_WAYLAND**: Ubuntu 23.04 Lunar with Gnome Shell and Wayland

- [Install build artifacts of this PR](https://github.com/keymanapp/keyman/wiki/How-to-test-artifacts-from-pull-requests-for-Keyman-for-Linux)

- Reboot
- Install Nanum fonts for typing Korean:

  ```bash
  sudo apt install fonts-nanum
  ```

- Change "Complex text layout" setting in LibreOffice Writer:

  <details>
  <summary>Expand to see the steps</summary>

  - In LO Tools/Options, go to the Language Settings/Languages tab. Under "Default Languages for Documents" check the box for "Asian" and select "Korean (RoK)" as language, and check "Complex text layout" and select "Khmer" as language.

    ![Screenshot from 2022-06-23 12-13-56](https://user-images.githubusercontent.com/181336/175276071-2fda6202-b2a2-498a-9638-517409e68dcb.png)

  - You'll also have to change the default style: go to Styles/Manage Styles, right-click on "Default Paragraph Style" and select Modify. Go to the Font tab. For Asian Text Font, select Family "NanumMyeongjo" and Language "Korean (RoK)". For CTL Font, select Family "Khmer Mondulkiri" (or another "Khmer" font) and Language "Khmer".

    ![Screenshot from 2022-06-23 12-11-35](https://user-images.githubusercontent.com/181336/175275721-d3b282e0-4efc-42b8-89aa-493468981141.png)
  </details>
</details>

 - Install "IPA (SIL)", "Korean KORDA Jamo (SIL)" , and "Khmer Angkor" keyman keyboards

## Tests

- **TEST_IPA_WRITER:** Open LibreOffice Writer. Switch to "IPA (SIL)" keyboard. Type `n>`. Verify that the result is "ŋ".
- **TEST_IPA_GEDIT:** Open gedit. Switch to "IPA (SIL)" keyboard. Type `n>`. Verify that the result is "ŋ".
- **TEST_KO_WRITER:** Open LibreOffice Writer. Switch to "Korean KORDA Jamo (SIL)" keyboard. Type `han<space>geul<space>`. Verify that the result is "한글".
- **TEST_KO_GEDIT:** Open gedit. Switch to "Korean KORDA Jamo (SIL)" keyboard. Type `han<space>geul<space>`. Verify that the result is "한글".
- **TEST_KM_WRITER:** Open LibreOffice Writer. Switch to "Khmer Angkor" keyboard. Type `xEjmr`. Verify that the output is "ខ្មែរ".
- **TEST_KM_GEDIT:** Open gedit. Switch to "Khmer Angkor" keyboard. Type `xEjmr`. Verify that the output is "ខ្មែរ".
- **TEST_CONTEXT:**
  - Open gedit
  - Switch to "IPA (SIL)" keyboard.
  - Type
    ```
    an
    m
    ```
  - click after `n`
  - type `>`
  - verify that the output is
    ```
    aŋ
    m
    ```
